charger/pulsatrix.go

```diff
@@ -0,0 +1,256 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"nhooyr.io/websocket"
+)
+
+// pulsatrix charger implementation
+type PulsatrixCharger struct {
+	conn        *websocket.Conn
+	ctx         context.Context
+	hostname    string
+	log         *util.Logger
+	tryRead     int
+	enState     bool
+	bo          *backoff.ExponentialBackOff
+	signaledAmp float64
+	mutex       sync.Mutex
+	WebsocketData
+}
+
+type WebsocketData struct {
+	VehicleStatus     string    `json:"vehicleStatus"`
+	LastActivePower   float64   `json:"lastActivePower"`
+	PhaseVoltage      []float64 `json:"voltage"`
+	PhaseAmperage     []float64 `json:"amperage"`
+	AllocatedAmperage float64   `json:"allocatedAmperage"`
+	EnergyImported    float64   `json:"energyImported"`
+}
+
+func init() {
+	registry.Add("pulsatrix", NewPulsatrixFromConfig)
+}
+
+// NewPulsatrixtFromConfig creates a pulsatrix charger from generic config
+func NewPulsatrixFromConfig(other map[string]interface{}) (api.Charger, error) {
+	var cc = struct {
+		Host string
+	}{}
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+	return NewPulsatrix(cc.Host)
+}
+
+// NewPulsatrix creates pulsatrix charger
+func NewPulsatrix(hostname string) (*PulsatrixCharger, error) {
+	wb := PulsatrixCharger{}
+	wb.log = util.NewLogger("pulsatrix")
+	wb.hostname = hostname
+	wb.bo = backoff.NewExponentialBackOff()
+	wb.bo.InitialInterval = 30 * time.Second
+	wb.bo.MaxInterval = 5 * time.Minute
+	return &wb, wb.connectWs(hostname)
+}
+
+// ConnectWs connects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) connectWs(hostname string) error {
+	uri := fmt.Sprintf("ws://%s/api/ws", hostname)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	c.ctx = ctx
+	conn, _, err := websocket.Dial(ctx, uri, nil)
+
+	if err != nil {
+		if strErr := fmt.Sprintf("%s", err); strErr == "websocket: bad handshake" {
+			c.log.ERROR.Println("bad handshake. Make sure the IP and port are correct")
+		} else {
+			c.log.ERROR.Println("error connecting to websocket:", err)
+			c.reconnectWs()
+		}
+		return err
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			c.log.ERROR.Println("Recovered from panic:", r)
+		}
+	}()
+
+	c.enState = false
+	c.conn = conn
+	c.handleError(c.Enable(false))
+	go c.wsReader()
+	go c.heartbeat()
+	return nil
+}
+
+// Heartbeat sends a heartbeat to the pulsatrix SECC
+func (c *PulsatrixCharger) heartbeat() {
+	for {
+		if c.enState {
+			c.handleError(c.Enable(true))
+			c.handleError(c.MaxCurrentMillis(c.signaledAmp))
+		} else {
+			c.handleError(c.Enable(false))
+		}
+		time.Sleep(3 * time.Minute)
+	}
+}
+
+func (c *PulsatrixCharger) handleError(err error) {
+	if err != nil {
+		c.log.ERROR.Println("error:", err)
+	}
+}
+
+// ReconnectWs reconnects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) reconnectWs() {
```

Comment:
The whole reconnect logic still feels much to complex for stable operation. It's not obvious to me if this is really synchronized or racy?